File: FactoryMethod.py
#Classe base Pessoa:
class Pessoa:
    def __init__(self, nome, nivel):
        self.nome = nome
        self.nivel =nivel
        
    def mostarPermissao(self):
        print(f"{self.nome} - Cargo: {self.__class__.__name__}, Permissão: Nível {self.nivel}")
        
#Subclasse:
class Aluno(Pessoa):
    def __init__(self, nome):
        super().__init__(nome, 0)
        
class Administrativo(Pessoa):
    def __init__(self, nome):
        super().__init__(nome, 1)
        
class Professor(Pessoa):
    def __init__(self, nome):
        super().__init__(nome, 2)
        
class Visitante(Pessoa):
    def __init__(self, nome):
        super().__init__(nome, 3)
        
#Factory
class FabricaPessoa:
    def criarPessoa(self, nome, cargo):
        if cargo.lower() == "aluno":
            return Aluno(nome)
        elif cargo.lower() == "administrativo":
            return Administrativo(nome)
        elif cargo.lower() == "professor":
            return Professor(nome)
        elif cargo.lower() == "visitante":
            return Visitante(nome)
        else:
            raise ValueError("Cargo Inválido!!")
        
#Uso
fabrica = FabricaPessoa()

p1 = fabrica.criarPessoa("Bruna" , "aluno")
p2 = fabrica.criarPessoa("Andrea" , "administrativo")
p3 = fabrica.criarPessoa("Vinicius", "professor")
p4 = fabrica.criarPessoa("Ana" , "visitante")

p1.mostarPermissao()
p2.mostarPermissao()
p3.mostarPermissao()
p4.mostarPermissao()

File: FactoryMethod.js
// Classe de pessoa
class Pessoa {
    constructor(nome, nivel){
        this.nome = nome;
        this.nivel = nivel;
    }
    mostrarPermissao(){
        console.log(`${this.nome} Cargo - ${this.constructor.name}, Permissão: Nível ${this.nivel}`);
    }
}

// Subclasse
class Aluno extends Pessoa{
    constructor(nome){
        super(nome, 0);
    }
}

class Administrativo extends Pessoa{
    constructor(nome){
        super(nome, 1);
    }
}

class Professor extends Pessoa{
    constructor(nome){
        super(nome, 2);
    }
}

class Visitante extends Pessoa{
    constructor(nome){
        super(nome, 3)
    }
}
// Factory
class fabricaPessoa{
    criarPessoa(nome, Cargo){
        switch(Cargo.toLowerCase()){
            case "aluno": return new Aluno(nome);
            case "administrativo": return new Administrativo(nome);
            case "professor": return new Professor(nome);
            case "visitante": return new Visitante(nome);
            default: throw new Error("Cargo Invalido!!")
        }
    }
}
// Uso
const fabrica = new fabricaPessoa();

const p1 = fabrica.criarPessoa("Bruna", "Aluno");
const p2 = fabrica.criarPessoa("Andrea", "Administrativo");
const p3 = fabrica.criarPessoa("Vinicius", "Professor");
const p4 = fabrica.criarPessoa("Ana", "Visitante");


p1.mostrarPermissao();
p2.mostrarPermissao();
p3.mostrarPermissao();
p4.mostrarPermissao();